### PR TITLE
[flutter_tools] add support for determinate progress bars in cli

### PR DESCRIPTION
**Pull request by jonahwilliams**
## Description

Add support for a determinate progress bar in logger.dart. Uses either unicode blocks or extended ascii, based on whether emoji support is detected. Use this new progress bar in the doctor checks.

This is intended to be a replacement for the status spinner for cases where we know some unit of completion (i.e. validators, downloaded bytes, et cetera)

ezgif-7-9dbe8a92dd6e

### CODE REVIEW

File: packages/flutter_tools/lib/src/base/logger.dart [hunk 1]
```diff
@@ -1042,3 +1071,197 @@ class AnsiStatus extends AnsiSpinner {
...
+  @override
+  void start(num maximum, String label, { String unit = '' }) {
+    if (_started) {
+      throw StateError('Progress cannot be started when it is already started');
+    }
+    _started = true;
+  }
+
+  @override
+  void update(num delta) { }
```
Comment (pedromassango) [inline packages/flutter_tools/lib/src/base/logger.dart:1120]: Is that necessary?

File: packages/flutter_tools/lib/src/base/logger.dart [hunk 2]
```diff
@@ -1042,3 +1071,197 @@ class AnsiStatus extends AnsiSpinner {
...
+///
+/// This is useful for machine modes where the progress output will not be
+/// surfaced to a user.
+class SilentProgress extends Progress {
+  bool _started = false;
+
+  @override
+  void finish() {
+    if (!_started) {
+      throw StateError('Progress cannot be finished when it has not been started');
```
Comment (pedromassango) [inline packages/flutter_tools/lib/src/base/logger.dart:1107]: We can end the message with Period/exclamation mark!

**Comment by pedromassango** [comment]
I would suggest it to be like in the example you shared previously:

Task Title...
Progress_bar $percentage $otherThings

File: packages/flutter_tools/lib/src/base/logger.dart [hunk 3]
```diff
@@ -1042,3 +1071,197 @@ class AnsiStatus extends AnsiSpinner {
...
+
+  String _formatPercentage(double percentage) {
+    final String value = (percentage * 100).toStringAsFixed(2);
+    if (percentage >= 1) {
+      return '$value %';
+    }
+    if (percentage < 0.1) {
+      return '  $value %';
+    }
+    return ' $value %';
```
Comment (pedromassango) [inline packages/flutter_tools/lib/src/base/logger.dart:1151]: Percentage symbol usually are placed next to the number

**Comment by pedromassango** [comment]
@jonahwilliams Why you closed this one?

**Comment by jonahwilliams** [comment]
don't have time to work on it this year

**Comment by pedromassango** [comment]
Isn’t this done yet?

**Pull request closed without merging**